tests/src/test/java/alluxio/client/fs/FileSystemS3UfsIntegrationTest.java

```diff
@@ -0,0 +1,116 @@
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.model.S3Object;
+import io.findify.s3mock.S3Mock;
+import org.apache.commons.io.IOUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+public class FileSystemS3UfsIntegrationTest extends BaseIntegrationTest {
+  private static final String TEST_CONTENT = "TestContents";
+  private static final String TEST_FILE = "test_file";
+  private static final int USER_QUOTA_UNIT_BYTES = 1000;
+  @Rule
+  public LocalAlluxioClusterResource mLocalAlluxioClusterResource =
+      new LocalAlluxioClusterResource.Builder()
+          .setProperty(PropertyKey.USER_FILE_BUFFER_BYTES, USER_QUOTA_UNIT_BYTES)
+          .setStartCluster(false)
+          .build();
+  private FileSystem mFileSystem = null;
+  private AmazonS3 mS3Client = null;
+  @Rule
+  public ExpectedException mThrown = ExpectedException.none();
+  private S3Mock mS3MockServer;
+  private static final String TEST_BUCKET = "test-bucket";
+
+  @Before
+  public void before() throws Exception {
+    mS3MockServer = new S3Mock.Builder().withPort(8001).withInMemoryBackend().build();
+    mS3MockServer.start();
+    AwsClientBuilder.EndpointConfiguration
+        endpoint = new AwsClientBuilder.EndpointConfiguration(
+        "http://localhost:8001", "us-west-2");
+    mS3Client = AmazonS3ClientBuilder
+        .standard()
+        .withPathStyleAccessEnabled(true)
+        .withEndpointConfiguration(endpoint)
+        .withCredentials(new AWSStaticCredentialsProvider(new AnonymousAWSCredentials()))
+        .build();
+    mS3Client.createBucket(TEST_BUCKET);
+
+    mLocalAlluxioClusterResource.setProperty(PropertyKey.UNDERFS_S3_ENDPOINT, "localhost:8001");
+    mLocalAlluxioClusterResource.setProperty(PropertyKey.UNDERFS_S3_ENDPOINT_REGION, "us-west-2");
+    mLocalAlluxioClusterResource.setProperty(PropertyKey.UNDERFS_S3_DISABLE_DNS_BUCKETS, true);
+    mLocalAlluxioClusterResource.setProperty(PropertyKey.MASTER_MOUNT_TABLE_ROOT_UFS,
+        "s3://" + TEST_BUCKET);
+    mLocalAlluxioClusterResource.start();
+    mFileSystem = mLocalAlluxioClusterResource.get().getClient();
+  }
+
+  @After
+  public void after() {
+    mS3Client = null;
+    try {
+      if (mS3MockServer != null) {
+        mS3MockServer.shutdown();
+      }
+    } finally {
+      mS3MockServer = null;
+    }
+  }
+
+  @Test
+  public void basicMetadataSync() throws IOException, AlluxioException {
+    mS3Client.putObject(TEST_BUCKET, TEST_FILE, TEST_CONTENT);
+    FileInStream fis = mFileSystem.openFile(new AlluxioURI("/" + TEST_FILE));
+    assertEquals(TEST_CONTENT, IOUtils.toString(fis, StandardCharsets.UTF_8));
+  }
+
+  @Test
+  public void basicWriteThrough() throws IOException, AlluxioException {
```

Comment:
what's the essential difference between this and FileSystemMasterS3UfsTest? why we need this two class to test the sync with ufs?

Reply:
FileSystemMasterS3UfsTest -> create a file system master with some mock components. you can access to the metadata of files only, but aren't able to read or write a file. -> more a unit test
FileSystemS3UfsintegrationTest -> create an alluxio clsuter that runs on local and do integration tests. you can do whatever you want to do as if this is a real alluxio cluster (it's real technically) -> an actual integration tests

Reply:
ok, got it